#188 - Adds network selection to contract address lookup

douglance opened the pull request:
I also added a throttle so we can optimistically return results on the first request. 

Closes #188

alex-miller-0 reviewed (commented):
I am without a Lattice currently and I'd like to test this one, but an unrelated comment:

 I'd prefer we not commit huge changes to `package-lock.json` in a feature PR (unless `package.json` is also updated with something needed by the feature). If you want to update packages can you please break that into a second PR?

douglance commented:
> I'd prefer we not commit huge changes to `package-lock.json` in a feature PR (unless `package.json` is also updated with something needed by the feature). If you want to update packages can you please break that into a second PR?

Yeah--that was not intentional. Not sure how it snuck in there. I reverted that change.

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+
+  // TEMPORARY FUNCTION TO REMOVE FUNCTIONS WITH ZERO LENGTH PARAM NAMES
+  // SEE: https://github.com/GridPlus/gridplus-sdk/issues/1540xfDcC959b0AA82E288E4154cB1C770C6c4e958a91
+  function TMP_REMOVE_ZERO_LEN_PARAMS(defs) {
```
You shouldn't need to do this. The linked issue has been resolved. Were you having issues without this function?

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  const getNetworkApiUrl = () =>
+    constants.CONTRACT_NETWORKS.find((cn) => cn.value === network).api;
```
Probably should fall back to default Ethereum url

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+  const [defs, setDefs] = useState([]);
+  const [network, setNetwork] = useState(defaultNetwork.value);
+
+  const resetState = () => {
```
This name is a bit confusing. At first read I thought you would be clearing out all state variables, including network. Perhaps something like `resetData`? Idk maybe there isn't a better name.

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+          // Map confusing error strings to better descriptions
+          if (resp.result === "Contract source code not verified") {
+            resp.result =
+              "Contract source code not published to Etherscan or not verified. Cannot determine data.";
```
You should add the name of the explorer ("Etherscan", "Polygonscan", etc) to `CONTRACT_NETWORKS` and replace `Etherscan` here with that variable.

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  const throttledFetch = useMemo(
+    () => throttle(fetchContractData, 5100),
```
Lol why `5100` ms? I'm guessing this is a typo.

douglance replied:
Because the throttle fires at exactly 5000 (according to the however it keeps time, which is probably not accurate of course), it sometimes triggers the scanning API to reject. I added a little buffer there to improve reliability.

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+  function addDefs() {
+    setInstalling(true);
+    setError("");
+    session.client.timeout = 2 * constants.ASYNC_SDK_TIMEOUT;
```
Please leave a comment explaining why you are switching this parameter twice

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,215 @@
+  const throttledFetch = useMemo(
+    () => throttle(fetchContractData, 5100),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [network]
```
What does this do?

douglance replied:
It recreates the `throttledFetch` function whenever the `network` variable changes, which in this case resets the throttle, so users aren't limited by the 5 second query against the new network. 

Because it resets the timer, users *can* switch networks, then quickly switch back and query and it will display the "Max rate limit reached, please use API Key for higher rate limit" message, but I don't think that will happen too often.

alex-miller-0 commented on src/util/helpers.js:
```diff
@@ -44,7 +44,28 @@ const constants = {
     RATE_LIMIT: 1000, // 1s between requests
     GET_ABI_URL: 'https://api.etherscan.io/api?module=contract&action=getabi&address=',
-}
+    CONTRACT_NETWORKS: [
```
See above comment and please add name of the explorer

alex-miller-0 reviewed (commented):
Looks mostly good. A few comments.

alex-miller-0 commented:
@douglance please add "Closes #188" to the PR description so it auto-closes the issue

alex-miller-0 commented on src/components/SearchCard.js:
```diff
@@ -0,0 +1,200 @@
+  );
+
+  const ErrorAlert = () => (
+    <Result status="error" title="Error" subTitle={error} />
```
I think this error screen would look better without a title, which creates a bunch of awkward white space. There's a big red X above it so I don't think the title is needed anyway.

![image](https://user-images.githubusercontent.com/7378490/147492032-e764d78a-9757-4ccc-92bd-9bed5a124f1f.png)

alex-miller-0 reviewed (commented):
Just one cosmetic request and then I'll approve. Also please set the issue number as closed in the PR description.

douglance commented:
> Just one cosmetic request and then I'll approve. Also please set the issue number as closed in the PR description.

Done and done.
